### PR TITLE
Return null when offset is null for lead/lag

### LINKED ISSUE #6413: Lead Function Semantics Diff: when offset is null, Presto return null, Velox return the default value supplied by user
### Bug description

Test case:
```java
// Presto TestLeadFunction.java
assertWindowQuery("lead(orderkey, null, -1) OVER (PARTITION BY orderstatus ORDER BY orderkey)",
        resultBuilder(TEST_SESSION, INTEGER, VARCHAR, INTEGER)
                .row(3, "F", null)
                .row(5, "F", null)
                .row(6, "F", null)
                .row(33, "F", null)
                .row(1, "O", null)
                .row(2, "O", null)
                .row(4, "O", null)
                .row(7, "O", null)
                .row(32, "O", null)
                .row(34, "O", null)
                .build());
```

The result of the window function above in Velox is `-1`.

### System information

Not Relevant.

### Relevant logs

_No response_

### PR DESCRIPTION
**Pull request by xumingming**
This is to match Presto's behavior for offset null scenario. Before this commit, Lead/Lag use kNullRow to describe both out of range and null offset scenario, in this commit we introduced another constant: kDefaultValueRow: we use kNullRow if offset is null, otherwise use kDefaultValueRow. We only copy default value from input vector for rows marked as kDefaultValueRow.

Fixes #6413

cc @mbasmanova @aditi-pandit

### CODE REVIEW

File: velox/functions/prestosql/window/LeadLag.cpp
Comment (aditi-pandit) [inline velox/functions/prestosql/window/LeadLag.cpp:90]: Nit - rewrite

Lead/Lag return default value (using kDefaultValueRow) if offsets for target rowNumbers are outside the partition. If offset is null, then the functions return null (using kNullRow) . 

kDefaultValueRow needs to be a negative number so that WindowPartition::extractColumn calls skip this row. It is set to -2 to distinguish it from kNullRow which is -1.

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/LeadLagTest.cpp:118]: Nit : Write full sentence ending in full-stop. "Out of range offsets return default value, whereas null offsets return null. "

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/LeadLagTest.cpp:289]: Add a comment // Default values.

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/LeadLagTest.cpp:293]: Add a comment // Null values.

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/LeadLagTest.cpp:297]: Add a comment "// Default values."

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (aditi-pandit) [inline velox/functions/prestosql/window/tests/LeadLagTest.cpp:301]: Add a comment "// Null values."

**Comment by aditi-pandit** [comment]
Thanks @xumingming for the PR. Minor nit comments only.

**Review by aditi-pandit — APPROVED** [review]
Thanks @xumingming

File: velox/functions/prestosql/window/LeadLag.cpp
Comment (mbasmanova) [inline velox/functions/prestosql/window/LeadLag.cpp:90]: Would you update documentation as well?

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (mbasmanova) [inline velox/functions/prestosql/window/tests/LeadLagTest.cpp:120]: My understanding is that here we test with default value being a constant 99. Can we also test with null default value and a non-constant default value that has a mix of nulls and non-nulls?

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (mbasmanova) [inline velox/functions/prestosql/window/tests/LeadLagTest.cpp:295]: maybe use nullEvery(7)

File: velox/functions/prestosql/window/LeadLag.cpp
Comment (xumingming) [reply velox/functions/prestosql/window/LeadLag.cpp:90]: done

File: velox/functions/prestosql/window/tests/LeadLagTest.cpp
Comment (xumingming) [reply velox/functions/prestosql/window/tests/LeadLagTest.cpp:122]: Added more tests.

File: velox/docs/functions/presto/window.rst
Comment (mbasmanova) [inline velox/docs/functions/presto/window.rst:139]: > if it is not specified ``null`` is returned.

What does "it" refer to here? Is it "offset"? Is so, what does it mean for offset to not be specified? You said above that default value for offset is 1. Please, clarify.

File: velox/docs/functions/presto/window.rst
Comment (mbasmanova) [inline velox/docs/functions/presto/window.rst:135]: I'm seeing that this PR changes the behavior in case when IGNORE NULLS is specific. Would you update this documentation to clarify what happens when IGNORE NULLS is on?

**Review by mbasmanova — COMMENTED** [review]
@xumingming James, would you clarify if there are test cases for IGNORE NULLS cases?

File: velox/docs/functions/presto/window.rst
Comment (xumingming) [reply velox/docs/functions/presto/window.rst:135]: by IGNORE NULLS do you mean the `if constexpr (ignoreNulls)` branch in `setRowNumbers` function, I have not touch that part in this PR

File: velox/functions/prestosql/window/LeadLag.cpp
Comment (mbasmanova) [inline velox/functions/prestosql/window/LeadLag.cpp:223]: @xumingming I believe this is the IGNORE NULLS code path. Since I see a change here, I assume this code path also changed.

File: velox/docs/functions/presto/window.rst
Comment (mbasmanova) [reply velox/docs/functions/presto/window.rst:135]: @aditi-pandit Aditi, it looks like we might be missing documentation for IGNORE NULLS clause: https://facebookincubator.github.io/velox/functions/presto/window.html

File: velox/functions/prestosql/window/LeadLag.cpp
Comment (xumingming) [reply velox/functions/prestosql/window/LeadLag.cpp:223]: En, indeed. I checked there is unit test for IGNORE NULLs path: https://github.com/facebookincubator/velox/blob/ead8a4ca3f9f369475559cc437f76a8afac253df/velox/functions/prestosql/window/tests/LeadLagTest.cpp#L391-L421

Comment (mbasmanova) [reply velox/functions/prestosql/window/LeadLag.cpp:223]: @xumingming Would you update the documentation to clarify the behavior when IGNORE NULLS is on and double check that all use cases with IGNORE NULLS are covered by tests?

File: velox/docs/functions/presto/window.rst
Comment (xumingming) [reply velox/docs/functions/presto/window.rst:135]: I'd suggest adding document for IGNORE NULL in a separate PR, seems many functions has `IGNORE NULLS` support(see [1]), so it might be better to describe IGNORE NULLS as a dedicated section.

[1]. https://duckdb.org/docs/sql/window_functions.html

Comment (mbasmanova) [reply velox/docs/functions/presto/window.rst:135]: Is the behavior of IGNORE NULLS  consistent across all functions? I assume lead / lag have their own semantics, no? Let's clarify what these are and make sure the code work properly.

Comment (xumingming) [reply velox/docs/functions/presto/window.rst:135]: Added some doc for IGNORE NULLs for lead/lag.

Comment (aditi-pandit) [reply velox/docs/functions/presto/window.rst:135]: > I'd suggest adding document for IGNORE NULL in a separate PR, seems many functions has `IGNORE NULLS` support(see [1]), so it might be better to describe IGNORE NULLS as a dedicated section.
> 
> [1]. https://duckdb.org/docs/sql/window_functions.html

@xumingming , @mbasmanova : I'll add doc for IGNORE NULLs in a later PR.

Comment (mbasmanova) [reply velox/docs/functions/presto/window.rst:135]: @aditi-pandit Thank you, Aditi. Can you help clarify IGNORE NULLS behavior for lead/lag so we can make sure the implementation is 100% correct?

Comment (aditi-pandit) [reply velox/docs/functions/presto/window.rst:135]: The meaning of IGNORE NULLs in lead/lag is as follows:

lead/lag need to get the column value of a row 'offset' rows away for the result. For IGNORE NULLs, we have to find the column value which is non-null 'offset' rows away. 

So the question for this PR is the code in this [comment ](https://github.com/facebookincubator/velox/pull/7254/files#r1376350944)... If there aren't non-null offset number of rows, then what should the result be ? As per Presto java it is the default value (if specified).

So this code is correct imo. 

But of course, its better to add a test. I think we can just add one more column in [ignoreNulls test](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/window/tests/LeadLagTest.cpp#L391) that has all null values only. By using that as a parameter for lead/lag should return default value for each row.

File: velox/functions/prestosql/window/LeadLag.cpp
Comment (aditi-pandit) [reply velox/functions/prestosql/window/LeadLag.cpp:223]: @xumingming , @mbasmanova : We should enhance the above ignoreNulls test to add a single column with only null values and use that as the column parameter in lead/lag to test the behavior in this change explicitly. Looking in the other tests for this scenario is fragile..

File: velox/docs/functions/presto/window.rst
Comment (xumingming) [reply velox/docs/functions/presto/window.rst:135]: Done, added all null values test.

File: velox/docs/functions/presto/window.rst
Comment (aditi-pandit) [inline velox/docs/functions/presto/window.rst:154]: Specify that default value is returned if rows without nulls are not found.

File: velox/docs/functions/presto/window.rst
Comment (aditi-pandit) [inline velox/docs/functions/presto/window.rst:157]: Nit : The scenario is more of non-null rows were not found vs all values are null

**Comment by xumingming** [comment]
Do we have further comments?

File: velox/docs/functions/presto/window.rst
Comment (aditi-pandit) [inline velox/docs/functions/presto/window.rst:157]: Nit : Rewrite "If not enough non-null values are found during offset counting, ...."

**Comment by aditi-pandit** [comment]
@xumingming : I'm satisified with the changes. Please get approval from @mbasmanova

**Comment by xumingming** [comment]
@mbasmanova Can you help merge this one?

**Comment by xumingming** [comment]
Gentle ping.

File: velox/docs/functions/presto/window.rst
Comment (mbasmanova) [inline velox/docs/functions/presto/window.rst:137]: > The offset can be any scalar expression.

This is not the case. This is true when user writes Presto SQL, but Velox  function accepts only constant value or a column.

File: velox/docs/functions/presto/window.rst
Comment (mbasmanova) [inline velox/docs/functions/presto/window.rst:133]: > [IGNORE NULLS]

We need to remove this since IGNORE NULLS is a SQL clause and not a function parameter.

File: velox/docs/functions/presto/window.rst
Comment (mbasmanova) [inline velox/docs/functions/presto/window.rst:142]: Functions do not allow suffixes. We need to find more accurate way to describe this. IGNORE NULLS can be mentioned at the top of this page as it applies to all window functions. Then, here we can say something like 

```
If IGNORE NULLS is specified, NULL values ...
```

**Review by mbasmanova — COMMENTED** [review]
Looks good % some nits.

@aditi-pandit Aditi, testing for window functions feels out of control. Let's invest in extending Fuzzer to cover window functions.

File: velox/docs/functions/presto/window.rst
Comment (xumingming) [inline velox/docs/functions/presto/window.rst:124]: This paragraph is from Presto's doc.

File: velox/docs/functions/presto/window.rst
Comment (xumingming) [inline velox/docs/functions/presto/window.rst:148]: Given that we have a paragraph at the top describing Ignore NULLs, the paragraph seems not needed any more?

**Comment by aditi-pandit** [comment]
> Looks good % some nits.
> 
> @aditi-pandit Aditi, testing for window functions feels out of control. Let's invest in extending Fuzzer to cover window functions.

@mbasmanova : Are you referring to Presto fuzzer testing or existing Velox fuzzer ? 

The main issues in Window testing have come about are when there is a choice of returning a default value or null and it seems like DuckDB behavior differs from Presto on those. For that we have added hard-coded Velox tests.

At the time being we have been adding tests in https://github.com/prestodb/presto/blob/master/presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java for window function queries. This has helped us discover many gaps. (There are couple pending PRs post fixes as well).

I'm inclined to build up that test in the short-term until fuzzer functionality is fleshed out. I hear that someone in Meta is going to work on it. I will stay on top of the fuzzer found issues for window functions.

**Comment by aditi-pandit** [comment]
@xumingming : Please add e2e Presto tests for this change in https://github.com/prestodb/presto/blob/master/presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java  as well.

**Comment by xumingming** [comment]
> @xumingming : Please add e2e Presto tests for this change in https://github.com/prestodb/presto/blob/master/presto-native-execution/src/test/java/com/facebook/presto/nativeworker/AbstractTestNativeWindowQueries.java as well.

I will. After this change is merged.

**Comment by xumingming** [comment]
@mbasmanova are you still not satisfied with the window document? If that is the case, how about merge this PR first and I(maybe @aditi-pandit can help here too) optimize the window doc in another PR?

**Review by mbasmanova — APPROVED** [review]
@xumingming Thank you for the fix. Looks good to me.

**Comment by mbasmanova** [comment]
@xumingming I'm getting "Diff Is Old or Base Revision Is Old" error. Would you rebase?

**Pull request closed without merging**
